examples/framesjs-starter/app/examples/next-routing-better/api.tsx

```diff
@@ -0,0 +1,388 @@
+    );
+  }
+}
+
+/**
+ * Creates a middleware that must be used in order to remove boilerplate from our frames pages
+ */
+export function createMiddleware<TFrameState extends FrameState = FrameState>(
+  initialState: TFrameState,
+  {
+    framesURL: framesBaseURL = "/frames",
+    framesHandlerURL = "/",
+  }: {
+    /**
+     * Relative path or absolute URL for POST request handler
+     *
+     * @default "/frames"
+     */
+    framesHandlerURL?: string;
+    /**
+     * Relative path or absolute URL which can be used as "basePath" for FrameLink routes
+     * @default "/"
+     */
+    framesURL?: string;
+  } = {}
+) {
+  return function middleware(request: NextRequest) {
+    const response = NextResponse.next();
+    const url = new URL(request.url);
+
+    response.headers.set(FRAMES_JS_SEARCH_PARAMS_HEADER_KEY, url.search);
+    response.headers.set(
+      FRAMES_JS_INITIAL_STATE_HEADER_KEY,
+      JSON.stringify(initialState)
+    );
+    response.headers.set(
+      FRAMES_JS_BASE_URL_HEADER_KEY,
+      resolveUrl(url, framesBaseURL).toString()
+    );
+    response.headers.set(
+      FRAMES_JS_HANDLER_URL_HEADER_KEY,
+      resolveUrl(url, framesHandlerURL).toString()
+    );
+
+    return response;
+  };
+}
+
+type GenerateLinkResult<TFrameState extends FrameState = FrameState> = {
+  path: string;
+  state: TFrameState;
+};
+
+type FramesFunctionResult<TFrameState extends FrameState = FrameState> = {
+  state: TFrameState;
+  previousFrame: PreviousFrame<TFrameState>;
+  generatePostButtonTarget(
+    to:
+      | string
+      | ((currentState: TFrameState) => GenerateLinkResult<TFrameState>)
+  ): string;
+  generatePostRedirectButtonTarget(buttonIndex: number, to: string): string;
+};
+
+// @todo in future version of react we can cache this function so it is computed only once per request. To use it now we would need to use canary version of react.
+export function frames<
+  TFrameState extends FrameState = FrameState,
+>(): FramesFunctionResult<TFrameState> {
+  const headerList = headers();
+
+  checkFrameContext();
+
+  // get search params from the headers that were provided by middleware
+  const searchParams = new URLSearchParams(
+    headerList.get(FRAMES_JS_SEARCH_PARAMS_HEADER_KEY) || ""
+  );
+  const initialState = JSON.parse(
+    headerList.get(FRAMES_JS_INITIAL_STATE_HEADER_KEY)!
+  ) as TFrameState;
+  const frameURL = headerList.get(FRAMES_JS_BASE_URL_HEADER_KEY)!;
+  const frameHandlerURL = headerList.get(FRAMES_JS_HANDLER_URL_HEADER_KEY)!;
+
+  const frame = getPreviousFrame<TFrameState>(
+    Object.fromEntries(searchParams.entries())
+  );
+
+  return {
+    state: frame.prevState ?? initialState,
+    previousFrame: frame,
+    generatePostButtonTarget(to) {
+      let path: string;
+      let state: TFrameState = frame.prevState ?? initialState;
+
+      if (typeof to === "string") {
+        path = to;
+      } else {
+        ({ path, state } = to(this.state));
```

Comment:
what if `to` includes mutations or side effects?

For example:
Press button, then try to mint an NFT (async), if success, render X, otherwise render Y. How does your solution solve this? It seems like this is the core confusion with the current library; that the "success/error" handling is always in the next render and thats a bit weird.

Reply:
The solution does not solve this. You can solve this in different route for example.